Fixes #26640 - Don't unregister existing hosts upon registration

jturel opened the pull request:
This PR fixes security issues with host registrations causing unregistration of other hosts due to how we match existing hosts by name or the dmi.system.uuid fact. The intention of the original code was to re-use the host record in Katello while unregistering & creating new registrations in backend Pulp and Candlepin.

**The change here completely removes the ability for a host to be unregistered by the registration of some other system even when they have the same hostname and dmi.system.uuid fact. We only reuse an existing host record when the hostname and dmi.system.uuid fact match the registering system *and* the matched host is not currently registered.**

To test:

- check out this PR
- register a client to your devel box (katello-client in Forklift makes it easy)
- note the dmi.system.uuid => `subscription-manager facts | grep dmi.system.uuid`
- trick the client into thinking it's unregistered: `subscription-manager clean` (the host profile remains in Katello)
- alter the dmi.system.uuid => `echo '{"dmi.system.uuid": "batmobile"}' > /etc/rhsm/facts/uuid.facts` (change can be verified by `subscription-manager facts`)
- attempt to register the host - the registration should fail:
```
root@katello-client vagrant]# subscription-manager register --username=admin --password=changeme --org=Default_Organization --environment=Library
Registering to: centos7-devel.jturel.example.com:443/rhsm
HTTP error (500 - Internal Server Error): Please remove hosts which match this host before registering: katello-client-a.jturel.example.com
```
- registration should also fail after restoring the original dmi.system.uuid and changing the hostname to something unique (since the UUID matches the first registration)

- Registration should be successful again only when the original host has been unregistered or removed from Katello (via API/UI/CLI) or the hostname *and* dmi.system.uuuid fact on the client are unique across all registered hosts

- The same steps can be followed when registering with an Activation Key `subscription-manager register --activationkey=MyKey --org=Default_Organization` and the results should be the same as described above

jturel commented:
@lhellebr this should address the issues you've identified with registrations causing unregistration of other hosts. Mind taking a look to see if it meets your expectations? It should apply cleanly to a downstream environment.

jturel commented:
[test katello]

lhellebr commented:
This is an improvement as in "safety" (issues caused by mistakes) but not as in "security" (issues caused intentionally). **Both** hostname and dmi.system.uuid can be spoofed, causing a host to get unregistered.

jturel commented:
@lhellebr good points - I have adjusted the code accordingly.

parthaa commented on app/models/katello/host/subscription_facet.rb:
```diff
@@ -242,33 +240,25 @@ def remove_subscriptions(pools_with_quantities)
+      def self.ensure_new_host(facts, organization)
         host_name = propose_existing_hostname(facts)
         uuid_fact_id = RhsmFactName.find_by(name: 'dmi::system::uuid')&.id || -1
         hosts = ::Host.unscoped.distinct.left_outer_joins(:fact_values)
```
Extra credit:  But can this be moved to a method in Host model via Host:::Managed extensions

jturel replied:
Might be a good idea. Let me think about that over lunch ;)

jturel replied:
I'd rather skip this as I don't see much benefit - but I do see why you would suggest it. Perhaps down the road a larger subset of subscription facet methods could be moved to Managed extensions since some do relate less to candlepin

parthaa commented:
> * alter the dmi.system.uuid => `echo '{"dmi.system.uuid": "batmobile"}' > /etc/rhsm/facts/uuid.facts` (change can be verified by `subscription-manager facts`)
> * attempt to register the host - the registration should fail:

Question for you. So it fails correctly but shouldnt the code be something other than 500. More like a 400 or 401
```
# subscription-manager register --org="Default_Organization" --activationkey="loca"
HTTP error (500 - Internal Server Error): Please remove hosts which match this host before registering: rhel7.partello.example.com
```

jturel commented:
> > * alter the dmi.system.uuid => `echo '{"dmi.system.uuid": "batmobile"}' > /etc/rhsm/facts/uuid.facts` (change can be verified by `subscription-manager facts`)
> > * attempt to register the host - the registration should fail:
> 
> Question for you. So it fails correctly but shouldnt the code be something other than 500. More like a 400 or 401
> 
> ```
> # subscription-manager register --org="Default_Organization" --activationkey="loca"
> HTTP error (500 - Internal Server Error): Please remove hosts which match this host before registering: rhel7.partello.example.com
> ```

I think "Bad Request" is referring to a malformed request - not a well-formed request that the server won't honor. Maybe I'm wrong?

parthaa commented:
> I think "Bad Request" is referring to a malformed request - not a well-formed request that the server won't honor. Maybe I'm wrong?

I wonder whether adding an exception here is a better idea. I need to figure out what's code for  standard error 

https://github.com/Katello/katello/blob/master/app/lib/katello/errors.rb

jturel commented:
@parthaa how about https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/422 ?

parthaa commented:
> @parthaa how about https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/422 ?

That sounds great :)

jturel commented:
@parthaa per our convo yesterday: what if I undid some of my changes here such that some of the original behavior of using an existing host profile can be achieved but only in these circumstances:

- subman registrations can only update an existing host when it is *not already registered* (it will not unregister it)
- the registering host must match in hostname and dmi.system.uuid otherwise a new profile will be created (or error if there are other hosts which match that are registered)

cc @lhellebr to stay in the loop on potential changes.

jturel commented:
@partha I updated the code per my last comment and also added the error scheme we talked about. I think this is a nice compromise between security and the original intended behavior. One thing I am not sure about is if we need this code in registration manager since we are only concerned with hosts that are already unregistered - but this might avoid errors?

https://github.com/Katello/katello/blob/a1c9280067607999cae43bab89b53ba870856b76/app/services/katello/registration_manager.rb#L37-L41

Let me know what you think about that
